common/datavalues/src/types/variant_value.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::str::FromStr;
+use std::cmp::Ordering;
+use std::fmt::Display;
+use std::fmt::Formatter;
+use std::ops::Deref;
+
+use common_exception::ErrorCode;
+use common_exception::Result;
+use serde_json::Value;
+
+#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, PartialEq, Eq)]
+pub struct VariantValue(Value);
+
+impl From<Value> for VariantValue {
+    fn from(val: Value) -> Self {
+        VariantValue(val)
+    }
+}
+
+impl From<&Value> for VariantValue {
+    fn from(val: &Value) -> Self {
+        VariantValue(val.clone())
+    }
+}
+
+impl AsRef<Value> for VariantValue {
+    fn as_ref(&self) -> &Value {
+        &self.0
+    }
+}
+
+impl Deref for VariantValue {
+    type Target = Value;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Default for VariantValue {
+    fn default() -> Self {
+        VariantValue::from(Value::Null)
+    }
+}
+
+impl FromStr for VariantValue {
+    type Err = ErrorCode;
+
+    fn from_str(s: &str) -> Result<Self> {
+        match serde_json::Value::from_str(s) {
+            Ok(v) => Ok(VariantValue::from(v)),
+            Err(_) => Err(ErrorCode::StrParseError(format!(
+                "Unknown VariantValue: {}",
+                s
+            ))),
+        }
+    }
+}
+
+// VariantValue compares as the following rule:
+// Null > Array > Object > String > Number > Boolean
```

Comment:
So we can give each struct a level value.

If the level is not equal, compare the level value.

Reply:
do you mean add a level filed in `struct VariantValue`?

Reply:
fn level(&self) -> u8 {
  match ...
}

Reply:
ok